File: backend/app/schemas/inventory.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.inventory import InventoryScopeType, InventoryStatus
from app.schemas.units import MaterialUnitOut


class InventorySessionCreate(BaseModel):
    scope_type: InventoryScopeType
    scope_ref_id: int | None = None

    @model_validator(mode="after")
    def _scope_ref_required_unless_warehouse(self):
        if self.scope_type != InventoryScopeType.WAREHOUSE and self.scope_ref_id is None:
            raise ValueError("scope_ref_id обязателен для scope_type != warehouse")
        return self


class InventorySessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    scope_type: InventoryScopeType
    scope_ref_id: int | None
    status: InventoryStatus
    started_by: int
    closed_by: int | None
    started_at: datetime
    closed_at: datetime | None
    expected_count: int
    scanned_count: int


class ScanRequest(BaseModel):
    location_code: str
    unit_id: int | None = None
    # Обязательны, только если unit_id не задан (единица не найдена — создание, как на приёмке).
    material: str | None = None
    color: str | None = None
    thickness: float | None = None
    manufacturer: str | None = None
    width_mm: float | None = Field(default=None, gt=0)
    length_m: float | None = Field(default=None, gt=0)

    @model_validator(mode="after")
    def _creation_fields_required_when_no_unit_id(self):
        if self.unit_id is None:
            missing = [
                name
                for name, value in (
                    ("material", self.material),
                    ("color", self.color),
                    ("thickness", self.thickness),
                    ("manufacturer", self.manufacturer),
                    ("width_mm", self.width_mm),
                    ("length_m", self.length_m),
                )
                if value is None
            ]
            if missing:
                raise ValueError(f"Единица не найдена — заполните для создания: {', '.join(missing)}")
        return self


class ScanResult(BaseModel):
    outcome: str  # confirmed / moved / surplus
    unit: MaterialUnitOut


class ShortageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    material_sku_id: int
    width_mm: float
    length_m: float
    location_code: str | None


class CloseSessionResult(BaseModel):
    session: InventorySessionOut
    confirmed_count: int
    moved_count: int
    surplus_count: int
    shortages: list[ShortageOut]


class ResolveShortageRequest(BaseModel):
    action: str  # "spisat" | "vernut_v_poisk"

File: frontend/src/api/inventory.ts
import { apiClient } from "./client";
import type { MaterialUnit } from "./units";

export type InventoryScopeType = "rack" | "warehouse" | "material_sku";
export type InventoryStatus = "in_progress" | "closed";

export interface InventorySession {
  id: number;
  scope_type: InventoryScopeType;
  scope_ref_id: number | null;
  status: InventoryStatus;
  started_by: number;
  closed_by: number | null;
  started_at: string;
  closed_at: string | null;
  expected_count: number;
  scanned_count: number;
}

export interface ScanRequest {
  location_code: string;
  unit_id?: number;
  material?: string;
  color?: string;
  thickness?: number;
  manufacturer?: string;
  width_mm?: number;
  length_m?: number;
}

export interface ScanResult {
  outcome: "confirmed" | "moved" | "surplus";
  unit: MaterialUnit;
}

export interface Shortage {
  id: number;
  material_sku_id: number;
  width_mm: number;
  length_m: number;
  location_code: string | null;
}

export interface CloseSessionResult {
  session: InventorySession;
  confirmed_count: number;
  moved_count: number;
  surplus_count: number;
  shortages: Shortage[];
}

export async function startSession(payload: { scope_type: InventoryScopeType; scope_ref_id?: number }): Promise<InventorySession> {
  const { data } = await apiClient.post<InventorySession>("/inventory-sessions", payload);
  return data;
}

export async function getSession(id: number): Promise<InventorySession> {
  const { data } = await apiClient.get<InventorySession>(`/inventory-sessions/${id}`);
  return data;
}

export async function scanUnit(sessionId: number, payload: ScanRequest): Promise<ScanResult> {
  const { data } = await apiClient.post<ScanResult>(`/inventory-sessions/${sessionId}/scan`, payload);
  return data;
}

export async function closeSession(sessionId: number): Promise<CloseSessionResult> {
  const { data } = await apiClient.post<CloseSessionResult>(`/inventory-sessions/${sessionId}/close`);
  return data;
}

export async function resolveShortage(
  sessionId: number,
  unitId: number,
  action: "spisat" | "vernut_v_poisk",
): Promise<InventorySession> {
  const { data } = await apiClient.post<InventorySession>(`/inventory-sessions/${sessionId}/resolve-shortage/${unitId}`, {
    action,
  });
  return data;
}
